columnflow/production/scale.py

```diff
@@ -0,0 +1,96 @@
+)
+def murmuf_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Producer that reads out mur and muf uncertainties on an event-by-event basis.
+    Can only be called with MC datasets.
+
+    Resources:
+
+       - https://cms-nanoaod-integration.web.cern.ch/integration/master/mc94X_doc.html
+    """
+
+    # stop here for data
+    if self.dataset_inst.is_data:
+        raise ValueError("attempt to read out mur/muf weights in data")
+
+    N_scaleweights = ak.num(events.LHEScaleWeight, axis=1)
+    if ak.any(N_scaleweights != 9):
+        N_scaleweights = N_scaleweights[N_scaleweights != 9]
+        raise Exception(
+            f"Number of LHEScaleWeights ({N_scaleweights}) is not as expected (9) "
+            f"in dataset {self.dataset_inst.name}",
+        )
+
+    if ak.any(events.LHEScaleWeight[:, 4] != 1):
+        raise Exception("The nominal LHEScaleWeight should always be 1")
+
+    events = set_ak_column(events, "mur_weight", events.LHEScaleWeight[:, 4])
+    events = set_ak_column(events, "mur_weight_up", events.LHEScaleWeight[:, 7])
+    events = set_ak_column(events, "mur_weight_down", events.LHEScaleWeight[:, 1])
+    events = set_ak_column(events, "muf_weight", events.LHEScaleWeight[:, 4])
+    events = set_ak_column(events, "muf_weight_up", events.LHEScaleWeight[:, 5])
+    events = set_ak_column(events, "muf_weight_down", events.LHEScaleWeight[:, 3])
+
+    return events
+
+
+@producer(
+    uses={"LHEScaleWeight"},
+    produces={"scale_weight", "scale_weight_up", "scale_weight_down"},
+)
+def murmuf_envelope_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Producer that determines the envelope of the mur/muf up and down variations on an event-by-event basis.
+    Can only be called with MC datasets.
+
+    Resources:
+
+       - https://cms-nanoaod-integration.web.cern.ch/integration/master/mc94X_doc.html
+    """
+
+    # stop here for data
+    if self.dataset_inst.is_data:
+        return events
+
+    N_scaleweights = ak.num(events.LHEScaleWeight, axis=1)
+    # For now, make an exception for st_schannel_had dataset; should be fixed with NanoAODv10
+    if ak.any(N_scaleweights != 9):
+        N_scaleweights = N_scaleweights[N_scaleweights != 9]
+        raise Exception(
+            f"Number of LHEScaleWeights ({N_scaleweights}) is not as expected (9) "
+            f"in dataset {self.dataset_inst.name}",
+        )
+
+    if ak.any(events.LHEScaleWeight[:, 4] != 1):
+        raise Exception("The nominal LHEScaleWeight should always be 1")
+
+    events = set_ak_column(events, "scale_weight", events.LHEScaleWeight[:, 4])
+
+    # for the up/down variations, take the max/min value of all possible combinations
+    # except mur=2, muf=0.5 (index 2) and mur=0.5, muf=2 (index 6) into account
+    idx_mask = (ak.local_index(events.LHEScaleWeight) != 2) & (ak.local_index(events.LHEScaleWeight) != 6)
+    considered_scale_weights = events.LHEScaleWeight[idx_mask]
```

Comment:
`local_index` can be quite expensive. Should ak arrays support advanced indexing with lists? In this case this would simply become

```python
considered_scale_weights = events.LHEScaleWeight[..., [0, 1, 3, 4, 5, 7, 8]]
```